# SPDX-License-Identifier: LGPL-3.0-only OR GPL-2.0-only OR GPL-3.0-only
# Copyright (C) 2026 OpenKaryon
#
# Qt 6.8.5 Windows 7: Restore DNS lookup through the pre-Windows-8 API.
#
# Derived from crystalidea/qt6windows7, commit
# 3df3aac758177b28843c4d1511a2bd7751fe82b6.
# Compared file by file with the pristine Qt 6.8.4 source archive; applies to 6.8.5.
# Applies from the root of qt-everywhere-src-6.8.5 with patch -p1.
diff --git a/qtbase/src/network/kernel/qdnslookup_win.cpp b/qtbase/src/network/kernel/qdnslookup_win.cpp
index 1b07776..5410eda 100644
--- a/qtbase/src/network/kernel/qdnslookup_win.cpp
+++ b/qtbase/src/network/kernel/qdnslookup_win.cpp
@@ -5,11 +5,9 @@
 #include <winsock2.h>
 #include "qdnslookup_p.h"
 
-#include <qendian.h>
+#include <qurl.h>
 #include <private/qnativesocketengine_p.h>
 #include <private/qsystemerror_p.h>
-#include <qurl.h>
-#include <qspan.h>
 
 #include <qt_windows.h>
 #include <windns.h>
@@ -65,100 +63,103 @@ DNS_STATUS WINAPI DnsQueryEx(PDNS_QUERY_REQUEST pQueryRequest,
 
 QT_BEGIN_NAMESPACE
 
-static DNS_STATUS sendAlternate(QDnsLookupRunnable *self, QDnsLookupReply *reply,
-                                PDNS_QUERY_REQUEST request, PDNS_QUERY_RESULT results)
-{
-    // WinDNS wants MTU - IP Header - UDP header for some reason, in spite
-    // of never needing that much
-    QVarLengthArray<unsigned char, 1472> query(1472);
-
-    auto dnsBuffer = new (query.data()) DNS_MESSAGE_BUFFER;
-    DWORD dnsBufferSize = query.size();
-    WORD xid = 0;
-    bool recursionDesired = true;
-
-    SetLastError(ERROR_SUCCESS);
-
-    // MinGW winheaders incorrectly declare the third parameter as LPWSTR
-    if (!DnsWriteQuestionToBuffer_W(dnsBuffer, &dnsBufferSize,
-                                    const_cast<LPWSTR>(request->QueryName), request->QueryType,
-                                    xid, recursionDesired)) {
-        // let's try reallocating
-        query.resize(dnsBufferSize);
-        if (!DnsWriteQuestionToBuffer_W(dnsBuffer, &dnsBufferSize,
-                                        const_cast<LPWSTR>(request->QueryName), request->QueryType,
-                                        xid, recursionDesired)) {
-            return GetLastError();
-        }
-    }
-
-    // set AD bit: we want to trust this server
-    dnsBuffer->MessageHead.AuthenticatedData = true;
-
-    QDnsLookupRunnable::ReplyBuffer replyBuffer;
-    if (!self->sendDnsOverTls(reply, { query.data(), qsizetype(dnsBufferSize) }, replyBuffer))
-        return DNS_STATUS(-1);   // error set in reply
-
-    // interpret the RCODE in the reply
-    auto response = reinterpret_cast<PDNS_MESSAGE_BUFFER>(replyBuffer.data());
-    DNS_HEADER *header = &response->MessageHead;
-    if (!header->IsResponse)
-        return DNS_ERROR_BAD_PACKET;        // not a reply
-
-    // Convert the byte order for the 16-bit quantities in the header, so
-    // DnsExtractRecordsFromMessage can parse the contents.
-    //header->Xid = qFromBigEndian(header->Xid);
-    header->QuestionCount = qFromBigEndian(header->QuestionCount);
-    header->AnswerCount = qFromBigEndian(header->AnswerCount);
-    header->NameServerCount = qFromBigEndian(header->NameServerCount);
-    header->AdditionalCount = qFromBigEndian(header->AdditionalCount);
-
-    results->QueryOptions = request->QueryOptions;
-    return DnsExtractRecordsFromMessage_W(response, replyBuffer.size(), &results->pQueryRecords);
-}
-
 void QDnsLookupRunnable::query(QDnsLookupReply *reply)
 {
-    // Perform DNS query.
-    alignas(DNS_ADDR_ARRAY) uchar dnsAddresses[sizeof(DNS_ADDR_ARRAY) + sizeof(DNS_ADDR)];
-    DNS_QUERY_REQUEST request = {};
-    request.Version = 1;
-    request.QueryName = reinterpret_cast<const wchar_t *>(requestName.constData());
-    request.QueryType = requestType;
-    request.QueryOptions = DNS_QUERY_STANDARD | DNS_QUERY_TREAT_AS_FQDN;
+    typedef BOOL (WINAPI *DnsQueryExFunc) (PDNS_QUERY_REQUEST, PDNS_QUERY_RESULT, PDNS_QUERY_CANCEL);
+    static DnsQueryExFunc myDnsQueryEx = 
+        (DnsQueryExFunc)::GetProcAddress(::GetModuleHandle(L"Dnsapi"), "DnsQueryEx");
+
+    PDNS_RECORD ptrStart = nullptr;
+
+    if (myDnsQueryEx)
+    {
+        // Perform DNS query.
+        alignas(DNS_ADDR_ARRAY) uchar dnsAddresses[sizeof(DNS_ADDR_ARRAY) + sizeof(DNS_ADDR)];
+        DNS_QUERY_REQUEST request = {};
+        request.Version = 1;
+        request.QueryName = reinterpret_cast<const wchar_t *>(requestName.constData());
+        request.QueryType = requestType;
+        request.QueryOptions = DNS_QUERY_STANDARD | DNS_QUERY_TREAT_AS_FQDN;
+
+        if (!nameserver.isNull()) {
+            memset(dnsAddresses, 0, sizeof(dnsAddresses));
+            request.pDnsServerList = new (dnsAddresses) DNS_ADDR_ARRAY;
+            auto addr = new (request.pDnsServerList->AddrArray) DNS_ADDR[1];
+            auto sa = new (addr[0].MaxSa) sockaddr;
+            request.pDnsServerList->MaxCount = sizeof(dnsAddresses);
+            request.pDnsServerList->AddrCount = 1;
+            // ### setting port 53 seems to cause some systems to fail
+            setSockaddr(sa, nameserver, port == DnsPort ? 0 : port);
+            request.pDnsServerList->Family = sa->sa_family;
+        }
 
-    if (protocol == QDnsLookup::Standard && !nameserver.isNull()) {
-        memset(dnsAddresses, 0, sizeof(dnsAddresses));
-        request.pDnsServerList = new (dnsAddresses) DNS_ADDR_ARRAY;
-        auto addr = new (request.pDnsServerList->AddrArray) DNS_ADDR[1];
-        auto sa = new (addr[0].MaxSa) sockaddr;
-        request.pDnsServerList->MaxCount = sizeof(dnsAddresses);
-        request.pDnsServerList->AddrCount = 1;
-        // ### setting port 53 seems to cause some systems to fail
-        setSockaddr(sa, nameserver, port == DnsPort ? 0 : port);
-        request.pDnsServerList->Family = sa->sa_family;
+        DNS_QUERY_RESULT results = {};
+        results.Version = 1;
+        const DNS_STATUS status = myDnsQueryEx(&request, &results, nullptr);
+        if (status >= DNS_ERROR_RCODE_FORMAT_ERROR && status <= DNS_ERROR_RCODE_LAST)
+            return reply->makeDnsRcodeError(status - DNS_ERROR_RCODE_FORMAT_ERROR + 1);
+        else if (status == ERROR_TIMEOUT)
+            return reply->makeTimeoutError();
+        else if (status != ERROR_SUCCESS)
+            return reply->makeResolverSystemError(status);
+
+        ptrStart = results.pQueryRecords;
     }
+    else
+    {
+        // Perform DNS query.
+        PDNS_RECORD dns_records = 0;
+        QByteArray requestNameUTF8 = requestName.toUtf8();
+        const QString requestNameUtf16 = QString::fromUtf8(requestNameUTF8.data(), requestNameUTF8.size());
+        IP4_ARRAY srvList;
+        memset(&srvList, 0, sizeof(IP4_ARRAY));
+        if (!nameserver.isNull()) {
+            if (nameserver.protocol() == QAbstractSocket::IPv4Protocol) {
+                // The below code is referenced from: http://support.microsoft.com/kb/831226
+                srvList.AddrCount = 1;
+                srvList.AddrArray[0] = htonl(nameserver.toIPv4Address());
+            } else if (nameserver.protocol() == QAbstractSocket::IPv6Protocol) {
+                // For supporting IPv6 nameserver addresses, we'll need to switch
+                // from DnsQuey() to DnsQueryEx() as it supports passing an IPv6
+                // address in the nameserver list
+                qWarning("%s", "IPv6 addresses for nameservers are currently not supported");
+                reply->error = QDnsLookup::ResolverError;
+                reply->errorString = tr("IPv6 addresses for nameservers are currently not supported");
+                return;
+            }
+        }
+        const DNS_STATUS status = DnsQuery_W(reinterpret_cast<const wchar_t*>(requestNameUtf16.utf16()), requestType, DNS_QUERY_STANDARD, &srvList, &dns_records, NULL);
+        switch (status) {
+        case ERROR_SUCCESS:
+            break;
+        case DNS_ERROR_RCODE_FORMAT_ERROR:
+            reply->error = QDnsLookup::InvalidRequestError;
+            reply->errorString = tr("Server could not process query");
+            return;
+        case DNS_ERROR_RCODE_SERVER_FAILURE:
+        case DNS_ERROR_RCODE_NOT_IMPLEMENTED:
+            reply->error = QDnsLookup::ServerFailureError;
+            reply->errorString = tr("Server failure");
+            return;
+        case DNS_ERROR_RCODE_NAME_ERROR:
+            reply->error = QDnsLookup::NotFoundError;
+            reply->errorString = tr("Non existent domain");
+            return;
+        case DNS_ERROR_RCODE_REFUSED:
+            reply->error = QDnsLookup::ServerRefusedError;
+            reply->errorString = tr("Server refused to answer");
+            return;
+        default:
+            reply->error = QDnsLookup::InvalidReplyError;
+            reply->errorString = QSystemError(status, QSystemError::NativeError).toString();
+            return;
+        }
 
-    DNS_QUERY_RESULT results = {};
-    results.Version = 1;
-    DNS_STATUS status = ERROR_INVALID_PARAMETER;
-    switch (protocol) {
-    case QDnsLookup::Standard:
-        status = DnsQueryEx(&request, &results, nullptr);
-        break;
-    case QDnsLookup::DnsOverTls:
-        status = sendAlternate(this, reply, &request, &results);
-        break;
+        ptrStart = dns_records;
     }
 
-    if (status == DNS_STATUS(-1))
-        return;     // error already set in reply
-    if (status >= DNS_ERROR_RCODE_FORMAT_ERROR && status <= DNS_ERROR_RCODE_LAST)
-        return reply->makeDnsRcodeError(status - DNS_ERROR_RCODE_FORMAT_ERROR + 1);
-    else if (status == ERROR_TIMEOUT)
-        return reply->makeTimeoutError();
-    else if (status != ERROR_SUCCESS)
-        return reply->makeResolverSystemError(status);
+    if (!ptrStart)
+        return;
 
     QStringView lastEncodedName;
     QString cachedDecodedName;
@@ -170,9 +171,9 @@ void QDnsLookupRunnable::query(QDnsLookupReply *reply)
     auto extractMaybeCachedHost = [&](QStringView name) -> const QString & {
         return lastEncodedName == name ? cachedDecodedName : extractAndCacheHost(name);
     };
-
+    
     // Extract results.
-    for (PDNS_RECORD ptr = results.pQueryRecords; ptr != NULL; ptr = ptr->pNext) {
+    for (PDNS_RECORD ptr = ptrStart; ptr != NULL; ptr = ptr->pNext) {
         // warning: always assign name to the record before calling extractXxxHost() again
         const QString &name = extractMaybeCachedHost(ptr->pName);
         if (ptr->wType == QDnsLookup::A) {
@@ -224,25 +225,6 @@ void QDnsLookupRunnable::query(QDnsLookupReply *reply)
             record.d->timeToLive = ptr->dwTtl;
             record.d->weight = ptr->Data.Srv.wWeight;
             reply->serviceRecords.append(record);
-        } else if (ptr->wType == QDnsLookup::TLSA) {
-            // Note: untested, because the DNS_RECORD reply appears to contain
-            // no records relating to TLSA. Maybe WinDNS filters them out of
-            // zones without DNSSEC.
-            QDnsTlsAssociationRecord record;
-            record.d->name = name;
-            record.d->timeToLive = ptr->dwTtl;
-
-            const auto &tlsa = ptr->Data.Tlsa;
-            const quint8 usage = tlsa.bCertUsage;
-            const quint8 selector = tlsa.bSelector;
-            const quint8 matchType = tlsa.bMatchingType;
-
-            record.d->usage = QDnsTlsAssociationRecord::CertificateUsage(usage);
-            record.d->selector = QDnsTlsAssociationRecord::Selector(selector);
-            record.d->matchType = QDnsTlsAssociationRecord::MatchingType(matchType);
-            record.d->value.assign(tlsa.bCertificateAssociationData,
-                                   tlsa.bCertificateAssociationData + tlsa.bCertificateAssociationDataLength);
-            reply->tlsAssociationRecords.append(std::move(record));
         } else if (ptr->wType == QDnsLookup::TXT) {
             QDnsTextRecord record;
             record.d->name = name;
@@ -254,7 +236,7 @@ void QDnsLookupRunnable::query(QDnsLookupReply *reply)
         }
     }
 
-    DnsRecordListFree(results.pQueryRecords, DnsFreeRecordList);
+    DnsRecordListFree(ptrStart, DnsFreeRecordList);
 }
 
 QT_END_NAMESPACE
